catchup/service.go

```diff
@@ -492,6 +512,65 @@ func (s *Service) sync() {
 	s.log.Infof("Catchup Service: finished catching up, now at round %v (previously %v). Total time catching up %v.", s.ledger.LastRound(), pr, elapsedTime)
 }
 
+// TODO this doesn't actually use the digest from cert!
+func (s *Service) fetchRound(cert agreement.Certificate, verifier *agreement.AsyncVoteVerifier) {
+	blockHash := bookkeeping.BlockHash(cert.Proposal.BlockDigest) // semantic digest (i.e., hash of the block header), not byte-for-byte digest
+	fetcher := s.latestRoundFetcherFactory.NewOverGossip(protocol.UniEnsBlockReqTag)
+	defer func() {
+		fetcher.Close()
+	}()
+	for s.ledger.LastRound() < cert.Round {
+		if fetcher.OutOfPeers(cert.Round) {
+			fetcher.Close()
+			// refresh peers and try again
+			logging.Base().Warn("fetchRound found no outgoing peers")
+			s.net.RequestConnectOutgoing(true, s.ctx.Done())
+			fetcher = s.latestRoundFetcherFactory.NewOverGossip(protocol.UniEnsBlockReqTag)
+		}
+		// Ask the fetcher to get the block somehow
+		block, fetchedCert, rpcc, err := s.innerFetch(fetcher, cert.Round)
+
+		if err != nil {
+			select {
+			case <-s.ctx.Done():
+				logging.Base().Debugf("fetchRound was asked to quit before we could acquire the block")
+				return
+			default:
+			}
+			logging.Base().Warnf("fetchRound could not acquire block, fetcher errored out: %v", err)
+			continue
+		}
+		rpcc.Close()
+
+		if block.Hash() == blockHash && block.ContentsMatchHeader() {
+			s.ledger.EnsureBlock(block, cert)
+			return
+		}
+		// Otherwise, fetcher gave us the wrong block
+		logging.Base().Warnf("fetcher gave us bad/wrong block (for round %d): fetched hash %v; want hash %v", cert.Round, block.Hash(), blockHash)
+
+		// As a failsafe, if the cert we fetched is valid but for the wrong block, panic as loudly as possible
+		if cert.Round == fetchedCert.Round &&
+			cert.Proposal.BlockDigest != fetchedCert.Proposal.BlockDigest &&
+			fetchedCert.Authenticate(*block, s.ledger, verifier) == nil {
+			s := "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
+			s += "!!!!!!!!!! FORK DETECTED !!!!!!!!!!!\n"
+			s += "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
+			s += "fetchRound called with a cert authenticating block with hash %v.\n"
+			s += "We fetched a valid cert authenticating a different block, %v. This indicates a fork.\n\n"
+			s += "Cert from our agreement service:\n%#v\n\n"
+			s += "Cert from the fetcher:\n%#v\n\n"
+			s += "Block from the fetcher:\n%#v\n\n"
+			s += "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
+			s += "!!!!!!!!!! FORK DETECTED !!!!!!!!!!!\n"
+			s += "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
+			s = fmt.Sprintf(s, cert.Proposal.BlockDigest, fetchedCert.Proposal.BlockDigest, cert, fetchedCert, block)
+			fmt.Println(s)
+			logging.Base().Error(s)
```

Comment:
Shouldn't this be a Panic? We don't want to keep running after a fork

Reply:
Current code doesn't panic. I'm not sure if panicking would be the smartest idea.
I think that logging to telemetry and having some forever-sleeping here would be a better thing to do. Keep in mind that the agreement is already locked-step with this one, so if we go to sleep here, both the agreement & catchup are disabled.
Either way, I would break that off to a separate change since it's a behavioral change.